src/main/java/org/traccar/database/OpenIdProvider.java

```diff
@@ -0,0 +1,170 @@
+
+import com.nimbusds.openid.connect.sdk.claims.UserInfo;
+
+public class OpenIdProvider {
+    private static final Logger LOGGER = LoggerFactory.getLogger(OpenIdProvider.class);
+    
+    public final Boolean force;
+    private final ClientID clientId;
+    private final ClientAuthentication clientAuth;
+    private URI callbackUrl;
+    private URI authUrl;
+    private URI tokenUrl;
+    private URI userInfoUrl;
+    private URI baseUrl;
+    private final String adminGroup;
+
+    private LoginService loginService;
+
+    @Inject
+    public OpenIdProvider(Config config, LoginService loginService) {     
+        this.loginService = loginService;
+
+        force = config.getBoolean(Keys.OPENID_FORCE);
+        clientId = new ClientID(config.getString(Keys.OPENID_CLIENTID));
+        clientAuth = new ClientSecretBasic(clientId, new Secret(config.getString(Keys.OPENID_CLIENTSECRET)));
+
+        try {
+            callbackUrl = new URI(config.getString(Keys.WEB_URL, "") + "/api/session/openid/callback");
+            authUrl = new URI(config.getString(Keys.OPENID_AUTHURL, ""));
+            tokenUrl = new URI(config.getString(Keys.OPENID_TOKENURL, ""));
+            userInfoUrl = new URI(config.getString(Keys.OPENID_USERINFOURL, ""));
+            baseUrl = new URI(config.getString(Keys.WEB_URL, ""));
+        } catch(URISyntaxException error) {
+            LOGGER.error("Invalid URIs provided in OpenID configuration");
+        }
+
+        adminGroup = config.getString(Keys.OPENID_ADMINGROUP);
+    }
+
+    public URI createAuthUri() {
+        AuthenticationRequest request = new AuthenticationRequest.Builder(
+                new ResponseType("code"),
+                new Scope("openid", "profile", "email", "groups"),
+                clientId,
+                callbackUrl)
+                .endpointURI(authUrl)
+                .state(new State())
+                .nonce(new Nonce())
+                .build();
+
+        return request.toURI();
+    }
+
+    private OIDCTokenResponse getToken(AuthorizationCode code) throws IOException, ParseException, GeneralSecurityException {
+        AuthorizationGrant codeGrant = new AuthorizationCodeGrant(code, callbackUrl);
+        TokenRequest tokenRequest = new TokenRequest(tokenUrl, clientAuth, codeGrant);
+
+        HTTPResponse tokenReq = tokenRequest.toHTTPRequest().send();
+        TokenResponse tokenResponse = OIDCTokenResponseParser.parse(tokenReq);
+        if (!tokenResponse.indicatesSuccess()) {
+            LOGGER.warn("Invalid authorization code provided to OpenID callback");
+            throw new GeneralSecurityException("Unable to authenticate with the OpenID Connect provider.");
+        }
+
+        return (OIDCTokenResponse) tokenResponse.toSuccessResponse();
+    }
+
+    private UserInfo getUserInfo(BearerAccessToken token) throws IOException, ParseException, GeneralSecurityException {
+        UserInfoResponse userInfoResponse;
```

Comment:
Why is this defined separately here? You can move it to the same line as the assignment.